AMDGPU: Custom lower fptrunc vectors for f32 -> f16

changpeng opened the pull request:
The latest asics support v_cvt_pk_f16_f32 instruction. However current implementation of vector fptrunc lowering fully scalarizes the vectors, and the scalar conversions may not always be combined to generate the packed one.
  We made v2f32 -> v2f16 legal in https://github.com/llvm/llvm-project/pull/139956. This work is an extension to handle wider vectors. Instead of fully scalarization, we split the vector to packs (v2f32 -> v2f16) to ensure the packed conversion can always been generated.

changpeng committed:
AMDGPU: Custom lower vector fptrunc of f32 -> f16

  GFx950+ supports v_cvt_pk_f16_f32. However current implementation
of vector fptrunc lowering fully scalarizes the vector, and the scalar
conversions may not always be combined to generate the packed one.

  We made v2f32 -> v2f16 legal in https://github.com/llvm/llvm-project/pull/139956.
This work is an extension to handle wider vectors. Instead of fully scalarization,
we split the vector to packs (v2f32 -> v2f16) to ensure the packed conversion can always
been generated.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -1061,9 +1061,10 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
-  if (ST.hasCvtPkF16F32Inst())
-    FPTruncActions.legalFor({{S32, S64}, {S16, S32}, {V2S16, V2S32}});
-  else
+  if (ST.hasCvtPkF16F32Inst()) {
+    FPTruncActions.legalFor({{S32, S64}, {S16, S32}, {V2S16, V2S32}})
+                  .customFor({{V4S16, V4S32}, {V8S16, V8S32}});
+  } else
     FPTruncActions.legalFor({{S32, S64}, {S16, S32}});
```
```suggestion
  if (ST.hasCvtPkF16F32Inst()) {
    FPTruncActions.legalFor({{S32, S64}, {S16, S32}, {V2S16, V2S32}})
                  .customFor({{V4S16, V4S32}, {V8S16, V8S32}});
  } else {
    FPTruncActions.legalFor({{S32, S64}, {S16, S32}});
  }
```

Also it looks like the format is off

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2749,6 +2752,20 @@ bool AMDGPULegalizerInfo::legalizeMinNumMaxNum(LegalizerHelper &Helper,
+  assert (EltTy == S16 && "Only handle vectors of half");
+
+  // Split vector to packs.
+  return Helper.fewerElementsVector(MI, 0, LLT::fixed_vector(2, EltTy)) ==
```
```suggestion
  return Helper.fewerElementsVector(MI, /*TypeIdx=*/0, LLT::fixed_vector(2, EltTy)) ==
```

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2749,6 +2752,20 @@ bool AMDGPULegalizerInfo::legalizeMinNumMaxNum(LegalizerHelper &Helper,
+                                          MachineRegisterInfo &MRI) const {
+  Register DstReg = MI.getOperand(0).getReg();
+  LLT DstTy = MRI.getType(DstReg);
+  assert (DstTy.isVector() && DstTy.getNumElements() > 2);
```
I think the format is off here

shiltian reviewed (commented):
looks good but please fix the clang-format issue

changpeng committed:
AMDGPU: Custom lower vector fptrunc of f32 -> f16

  GFx950+ supports v_cvt_pk_f16_f32. However current implementation
of vector fptrunc lowering fully scalarizes the vector, and the scalar
conversions may not always be combined to generate the packed one.
  We made v2f32 -> v2f16 legal in https://github.com/llvm/llvm-project/pull/139956.
This work is an extension to handle wider vectors. Instead of fully scalarization,
we split the vector to packs (v2f32 -> v2f16) to ensure the packed conversion can always
been generated.

 Fix clang-format

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2749,6 +2753,21 @@ bool AMDGPULegalizerInfo::legalizeMinNumMaxNum(LegalizerHelper &Helper,
   return Helper.lowerFMinNumMaxNum(MI) == LegalizerHelper::Legalized;
 }
 
+bool AMDGPULegalizerInfo::legalizeFPTrunc(LegalizerHelper &Helper,
+                                          MachineInstr &MI,
+                                          MachineRegisterInfo &MRI) const {
+  Register DstReg = MI.getOperand(0).getReg();
+  LLT DstTy = MRI.getType(DstReg);
+  assert(DstTy.isVector() && DstTy.getNumElements() > 2);
+  LLT EltTy = DstTy.getElementType();
+  assert(EltTy == S16 && "Only handle vectors of half");
+
+  // Split vector to packs.
+  LLT PkTy = LLT::fixed_vector(2, EltTy);
+  return Helper.fewerElementsVector(MI, /*TypeIdx=*/0, PkTy) ==
+         LegalizerHelper::Legalized;
+}
+
```
This is just ordinary fewerElementsVector, this doesn't need t o be custom

changpeng replied:
> You should just need `.clampMaxNumElements(0, S16, 2)`

This looks good. I was trying to use .fewerElementsIf to exclude non-power-of-2 cases, but it should be fine there.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6900,14 +6902,44 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG, SDValue Op,
+    SDValue PkSrc = DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, SrcPkVT, Src,
+                                DAG.getConstant(Index, DL, MVT::i32));
+    SDValue PkDst = DAG.getNode(Opc, DL, DstPkVT, PkSrc,
+                                DAG.getTargetConstant(0, DL, MVT::i32));
```
This should preserve the original value for the round flag

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6900,14 +6902,44 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG, SDValue Op,
                            DAG.getTargetConstant(0, DL, MVT::i32));
 }
 
+SDValue SITargetLowering::SplitFP_ROUNDVectorToPacks(SDValue Op,
```
This would be shorter if it followed splitUnaryVectorOp, except this one needs to carry forward the rounding flag operand

arsenm replied:
splitUnaryVectorOp could be adjusted to treat the input and output types differently, but the real reason this case is  different is the extra rounding flag operand.

I meant you should create a new helper function, which looks like splitUnaryOp. That is, use DAG.SplitVectorOperand instead of manually splitting

changpeng replied:
Okay

arsenm commented on llvm/test/CodeGen/AMDGPU/fptrunc.v2f16.no.fast.math.ll:
```diff
@@ -12,18 +12,124 @@ define <2 x half> @v_test_cvt_v2f32_v2f16(<2 x float> %src) {
+  %res = fptrunc <4 x float> %src to <4 x half>
+  ret <4 x half> %res
+}
+
```
Need to test the 3 x cases

changpeng replied:
> 3x is not marked as custom lowering in this PR

+    addRegisterClass(MVT::v3f16, &AMDGPU::SReg_64RegClass);

  Have to add the above line if we want to custom lower v3f16. I am afraid this will affect cases that do not custom lower v3f16 without making v3f16 legal. So it is better we do not custom lower v3f16 for fptrunc of v32->f16 now (which is just an optimization to generate v_cvt_pk).

arsenm replied:
It doesn't matter if it's added as custom lowered or not in this PR, it should be tested. You probably do not need to do anything to handle it correctly, but it needs the test.

You also do not need to make the type legal to custom lower, custom lowering on illegal types calls ReplaceNodeResults

arsenm commented on llvm/test/CodeGen/AMDGPU/fptrunc.v2f16.no.fast.math.ll:
```diff
@@ -12,18 +12,124 @@ define <2 x half> @v_test_cvt_v2f32_v2f16(<2 x float> %src) {
+  %sum5 = fadd half %sum2, %sum3
+  %rslt = fadd half %sum4, %sum5
+  ret half %rslt
 }
```
Test 16 x

changpeng committed:
AMDGPU: Custom lower vector fptrunc of f32 -> f16

  GFx950+ supports v_cvt_pk_f16_f32. However current implementation
of vector fptrunc lowering fully scalarizes the vector, and the scalar
conversions may not always be combined to generate the packed one.
  We made v2f32 -> v2f16 legal in https://github.com/llvm/llvm-project/pull/139956.
This work is an extension to handle wider vectors. Instead of fully scalarization,
we split the vector to packs (v2f32 -> v2f16) to ensure the packed conversion can always
been generated.

NOTE:
 1) re-implement the helper function the sam way as splitUnaryVectorOp;
 2) Use fewerElementsIf instead of custom lowering on the global isel path
 3) add checks for 3x and 16x vectors

changpeng committed:
AMDGPU: Custom lower vector fptrunc of f32 -> f16

  GFx950+ supports v_cvt_pk_f16_f32. However current implementation
of vector fptrunc lowering fully scalarizes the vector, and the scalar
conversions may not always be combined to generate the packed one.
  We made v2f32 -> v2f16 legal in https://github.com/llvm/llvm-project/pull/139956.
This work is an extension to handle wider vectors. Instead of fully scalarization,
we split the vector to packs (v2f32 -> v2f16) to ensure the packed conversion can always
been generated.

NOTE:
     minor changes

changpeng committed:
AMDGPU: Custom lower vector fptrunc of f32 -> f16

  GFx950+ supports v_cvt_pk_f16_f32. However current implementation
of vector fptrunc lowering fully scalarizes the vector, and the scalar
conversions may not always be combined to generate the packed one.
  We made v2f32 -> v2f16 legal in https://github.com/llvm/llvm-project/pull/139956.
This work is an extension to handle wider vectors. Instead of fully scalarization,
we split the vector to packs (v2f32 -> v2f16) to ensure the packed conversion can always
been generated.

NOTE: Use .clampMaxNumElements(0, S16, 2)

changpeng commented:
Any further suggestions/comments?  Thanks.

changpeng commented:
ping!

changpeng committed:
AMDGPU: Custom lower fptrunc vectors for f32 -> f16

 clang-format fixes

changpeng committed:
AMDGPU: Custom lower fptrunc vectors for f32 -> f16

 clang format fix

shiltian commented on llvm/lib/Target/AMDGPU/SIISelLowering.h:
```diff
@@ -145,6 +145,7 @@ class SITargetLowering final : public AMDGPUTargetLowering {
 
   /// Custom lowering for ISD::FP_ROUND for MVT::f16.
   SDValue lowerFP_ROUND(SDValue Op, SelectionDAG &DAG) const;
+  SDValue splitFP_ROUNDVectorOp(SDValue Op, SelectionDAG &DAG) const;
```
Did I miss anything here? This function was added but not implemented?

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6900,14 +6903,35 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG, SDValue Op,
   EVT DstVT = Op.getValueType();
 
-  if (DstVT == MVT::v2f16) {
+  if (DstVT.isVector() && DstVT.getScalarType() == MVT::f16) {
```
In a follow up can you look into extending this for v2bf16? I'm guessing in the ultimate expansion sequence, this will give a benefit even if the underlying v2 opcode isn't legal

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -1061,10 +1061,12 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
-  if (ST.hasCvtPkF16F32Inst())
-    FPTruncActions.legalFor({{S32, S64}, {S16, S32}, {V2S16, V2S32}});
-  else
+  if (ST.hasCvtPkF16F32Inst()) {
+    FPTruncActions.legalFor({{S32, S64}, {S16, S32}, {V2S16, V2S32}})
+        .clampMaxNumElements(0, S16, 2);
+  } else {
```
Don't think this is actually tested

arsenm reviewed (approved):
LGTM except I'm not sure globalisel is actually tested
